src/cpp/rtps/history/TopicPayloadPool_impl/DynamicReusable.hpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file DynamicReusable.hpp
+ */
+
+#ifndef RTPS_HISTORY_TOPICPAYLOADPOOLIMPL_DYNAMIC_REUSABLE_HPP
+#define RTPS_HISTORY_TOPICPAYLOADPOOLIMPL_DYNAMIC_REUSABLE_HPP
+
+#include <rtps/history/TopicPayloadPool.hpp>
+
+namespace eprosima {
+namespace fastrtps {
+namespace rtps {
+
+class DynamicReusableTopicPayloadPool : public TopicPayloadPool
+{
+    bool get_payload(
+            uint32_t size,
+            CacheChange_t& cache_change) override
+    {
+        if (!TopicPayloadPool::get_payload(size, cache_change))
+        {
+            return false;
+        }
+
+        // Resize if needed
+        if (size > cache_change.serializedPayload.max_size)
+        {
+            if (!resize_payload(cache_change.serializedPayload.data, cache_change.serializedPayload.max_size, size))
+            {
+                logError(RTPS_HISTORY, "Failed to resize the payload");
+                release_payload(cache_change);
+                return false;
+            }
+        }
```

Comment:
I'm concerned with context switching here, as `get_payload` will take and release the mutex, and `resize_payload` will do that again. 

As the code here is exactly the same as in `PreallocatedWithRealloc`, I would add a protected method to the base class that grows the buffer before releasing the mutex